packages/opentelemetry-types/src/metrics/metric.ts

```diff
@@ -0,0 +1,63 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Attributes } from '../trace/attributes';
+import { Resource } from '../resources/Resource';
+
+export interface MetricOptions {
+  // Description of the Metric.
+  description?: string;
+
+  // Unit of the Metric values.
+  unit?: string;
+
+  // List of attribute keys with dynamic values. Order of list is important
+  // as the same order must be used when supplying values for these attributes.
+  dynamicAttributes?: string[];
+
+  // List of attributes with constant values.
+  constantAttributes?: Attributes;
+
+  // Resource the metric is associated with.
+  resource?: Resource;
+
+  // Name of the component that reports the metric.
+  component?: string;
+}
+
+// Metric represents a base class for different types of metric preaggregations.
+export interface Metric<T> {
```

Comment:
Are there any bounds on what type `T` can be? And what does type `T` here mean?

Reply:
Sadly it's not really spelled out in the specification but there's a somewhat hidden type called `TimeSeries`. So let's say you have a counter metric. In order to actually add something to the metric, you need to get the `TimeSeries` first by calling `requestCount.getOrCreateTimeseries(customer_id)`. T represents the specific timeseries implementation.

Right now Counter and Gauge have the same interface in the specification (although maybe they shouldn't). I'm curious to see what the histogram and summary types are going to look like.

Reply:
I realize this is already merged, but getting caught up here. When I see a templated type like `T`, I tend to assume that as a user I can change that type.

But in this case it's really more like `unknown` - something opaque that just gets passed around. Why is the template needed on the `Metric` type? Could the implementation just be a private property or similar?